File: Problem1_Mjolnir/1_claude_sol2.java
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer st = new StreamTokenizer(br);

        st.nextToken();
        int T = (int) st.nval;
        StringBuilder sb = new StringBuilder();

        for (int t = 0; t < T; t++) {
            st.nextToken();
            int N = (int) st.nval;
            st.nextToken();
            long R = (long) st.nval;
            st.nextToken();
            long sCap = (long) st.nval;
            st.nextToken();
            long wCap = (long) st.nval;

            int[] impact = new int[N];
            int[] cost = new int[N];

            for (int i = 0; i < N; i++) {
                st.nextToken();
                long s = (long) st.nval;
                st.nextToken();
                long w = (long) st.nval;
                st.nextToken();
                int c = (int) st.nval;

                long im = s - w;
                if (im < 0)
                    im = 0;
                impact[i] = (int) im;
                cost[i] = c;
            }

            long capTotal = sCap + wCap;
            if (R > capTotal) {
                sb.append(-1).append('\n');
                continue;
            }

            int allowed = (int) (capTotal - R);
            final long INF = Long.MAX_VALUE / 4;

            // dp[w] = minimum convincing cost so that avengers left
            // un-convinced have combined impact <= w
            long[] dp = new long[allowed + 1];

            for (int i = 0; i < N; i++) {
                int im = impact[i];
                int c = cost[i];
                for (int w = allowed; w >= 0; w--) {
                    long convinceOpt = dp[w] + c;
                    long keepOpt = (w >= im) ? dp[w - im] : INF;
                    dp[w] = Math.min(convinceOpt, keepOpt);
                }
            }

            sb.append(dp[allowed]).append('\n');
        }

        System.out.print(sb);
    }
}

File: Problem1_Mjolnir/1_claude_sol2.py
import sys


def solve():
    data = sys.stdin.buffer.read().split()
    pos = 0

    def read_int():
        nonlocal pos
        v = data[pos]
        pos += 1
        return int(v)

    T = read_int()
    results = []
    INF = float('inf')

    for _ in range(T):
        N = read_int()
        R = read_int()
        s_cap = read_int()
        w_cap = read_int()

        impact_list = []
        cost_list = []

        for _ in range(N):
            s = read_int()
            w = read_int()
            c = read_int()
            im = s - w
            if im < 0:
                im = 0
            impact_list.append(im)
            cost_list.append(c)

        cap_total = s_cap + w_cap

        if R > cap_total:
            results.append(-1)
            continue

        allowed = cap_total - R

        # dp[w] = minimum total convincing cost such that the impact of
        # avengers left un-convinced does not exceed w
        dp = [0] * (allowed + 1)

        for im, c in zip(impact_list, cost_list):
            for w in range(allowed, -1, -1):
                convince_opt = dp[w] + c
                keep_opt = dp[w - im] if w >= im else INF
                dp[w] = convince_opt if convince_opt < keep_opt else keep_opt

        results.append(dp[allowed])

    sys.stdout.write("\n".join(map(str, results)) + "\n")


solve()
